File: Python/sibsort.py
import math

def sibsort(array):
	gap = math.floor(len(array) * .75) #											gap will determine how many indexes apart the comparisons happen
	while gap > 1:

		for index in range(0, len(array) - gap, 2): #									compare and swap gapped odd-even pairs
			if array[index] > array[index + gap]:
				array[index], array[index + gap] = array[index + gap], array[index]
		for index in range(1, len(array) - gap, 2): # 									compare and swap gapped even-odd pairs
			if array[index] > array[index + gap]:
				array[index], array[index + gap] = array[index + gap], array[index]
		gap = math.floor(gap * .75)
		if gap % 2 == 0: #												keep gap odd
			gap = gap - 1
	insertion(array) #													once gap is 1 or lower, call insertion to finish

# build a growing sorted list by inserting each piece encountered where it fits in that list
def insertion(array): 
	for index in range(1, len(array)):
		pointer = index - 1
		index_value = array[index]
		while index_value < array[pointer]:
			array[pointer], array[pointer + 1] = array[pointer + 1], array[pointer]
			if pointer > 0:
				pointer -= 1

File: Java/sibsort.java
public static int[] SibSort(int[] array) //					Change datatype of array for your use case!
{
	int gap = (int) (array.length * 0.75);
	if (gap % 2 == 0)
	{
		gap--;
	}
	while (gap > 1)
	{
		System.out.println(gap);
		for (int i = 0; i + gap < array.length; i += 2)
		{
			if (array[i] > array[i + gap])
			{
				array = Swap(array, i, i + gap);
			}
		}
		for (int i = 1; i + gap < array.length; i += 2)
		{
			if (array[i] > array[i + gap])
			{
				array = Swap(array, i, i + gap);
			}
		}
		gap = (int) (gap * 0.75);
		if (gap % 2 == 0)
		{
			gap--;
		}
	}
	array = Insertion(array);
	return array;
}

public static int[] Insertion(int[] array) //					Change datatype of array for your use case!
{
	for (int i = 1; i < array.length; i++)
	{
		int iValue = array[i];
		for (int j = i - 1; j >= 0 && array[j] > iValue; j--)
		{
			array = Swap(array, j, j + 1);
		}
	}
	return array;
}

public static int[] Swap(int[] array, int a, int b)
{
	int temp = array[a];
	array[a] = array[b];
	array[b] = temp;
	return array;
}
// by siblinghood
